### PR TITLE
RavenDB-12502 Graph query staleness

### PR DESCRIPTION
**Pull request by talweiss1982**
* also includes a fix for the Collection Destination optimization

### CODE REVIEW

File: src/Raven.Server/Documents/Queries/Graph/GraphQueryIndexGatherer.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+namespace Raven.Server.Documents.Queries.Graph
+{
+    public class GraphQueryIndexNamesGatherer: QueryPlanVisitor
+    {
+        public List<string> Indexes { get; set; } = new List<string>();
+
+        public override void VisitQueryQueryStep(QueryQueryStep qqs)
+        {
+            var name = qqs.GetIndexName;
+            if(name.IsNullOrWhiteSpace() == false)
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/Graph/GraphQueryIndexGatherer.cs:16]: I'm assuming that this filters out the collection queries?

File: src/Raven.Server/Documents/Queries/Graph/GraphQueryPlan.cs [hunk 2]
```diff
@@ -196,5 +199,87 @@ public List<Match> Execute()
...
+                {
+                    //we might get here if all indexes have timedout 
+                    return indexTimedout;
+                }
+            }
+        }
+
+        private List<Index> GetStaleIndexes(List<Index> indexes, long etag)
+        {
+            var staleList = new List<Index>();
```
Comment (ayende) [inline src/Raven.Server/Documents/Queries/Graph/GraphQueryPlan.cs:272]: Why not re-use the same instance? I don't see it being used for other stuff?

Comment (talweiss1982) [reply src/Raven.Server/Documents/Queries/Graph/GraphQueryPlan.cs:272]: I can't modify the list while iterating on it, so I need a 'to delete' list anyway so I might as well just create a new one.

File: src/Raven.Server/Documents/Queries/Graph/GraphQueryIndexGatherer.cs (hunk 1 above)
Comment (talweiss1982) [reply src/Raven.Server/Documents/Queries/Graph/GraphQueryIndexGatherer.cs:16]: I'm not sure I need to verify this.

Comment (talweiss1982) [reply src/Raven.Server/Documents/Queries/Graph/GraphQueryIndexGatherer.cs:16]: yes a collection query will return null

**Pull request merged**